laurent22/joplin
# Desktop: Resolves #4251: Refactor sidebar to better handle thousands of tags and notebooks

# Summary

This pull request refactors the desktop sidebar to:
1. Avoid re-generating the list of tags and folders when the selection changes.
2. Render the currently visible notes and/or tags, rather than all notes and tags on each re-render.
3. Have Playwright tests.
    - To support this change, several items that were missing accessibility labels have been them. This includes the "new note" button and the expand/unexpand folder buttons.

With this pull request,
- Adding tags is faster.
- Switching notebooks is faster.
- Moving a large notebook with thousands of items to the trash is still slow.
- Expanding a large notebook with many nested subnotebooks is still slow (though doesn't seem to be as slow as before).

Fixes #4251.

# Notes

- This should only affect desktop. A similar pull request may need to be created to refactor the mobile sidebar.
- At present, this pull request assumes that all notebook, tag, and header items have the same height in the sidebar.
   - At present, this height is `30px`.
   - This may not be the case if the user installs a theme like the [ReMoods theme](https://joplinapp.org/plugins/plugin/joplin.plugin.remoods.theme/) or has changed Joplin's notebook list styles with custom CSS.
- Much of the logic that was originally in `Sidebar.tsx` has been moved to the `useOnRenderItem.ts` hook.
- The following issues seem to have been present before this PR and are still present:
  - When dragging one folder to another, the note count for the parent folder may briefly go negative.
  - When adding a new notebook, the notebook is initially added below the trash folder, then moved to the correct location. This seems to be because we set `doRefreshFolders` to `true` instead of `now` [here](https://github.com/laurent22/joplin/blob/1ef4e574b7d6b8a40cc6c8d923bd8c4cf70ba8d2/packages/lib/BaseApplication.ts#L453) (`true` causes a delayed update, while `now` updates without delay).


# To-do

- [x] Test manually: How much faster is it?
   - Potentially add more: Include tests for note count and adding/removing tags.
-  **Follow-up pull request**: Make finding which notebooks are children of the current notebook faster. At present, it seems to be in $\mathcal{O}(n^2)$.
    - As a result, Joplin is still slow with many deeply-nested notebooks.

# Manual testing

On Ubuntu 23.10, I have manually verified that:
1. Dragging a note to a notebook heading moves that note to the target notebook.
2. Clicking on the "NOTEBOOKS" header hides/shows all notebooks
3. Clicking on the "TAGS" header hides/shows all tags
4. Selecting a notebook and/or tag outside of the viewport with the up/down arrow keys scrolls the viewport
5. Note counts update when dragging one notebook to another
6. It is still possible to use "go to anything" to search and scroll through search results with arrow keys
    - GoToAnything uses `ItemList`, which is refactored by this pull request.
7. On mobile (Android 10 emulator), with 4-5 notebooks, it is still possible to select and expand/collapse notebooks in the sidemenu.


Note that this pull request has automated tests, which additionally tests:
- Expanding/collapsing notebooks
- Dragging one notebook to another
- Dragging one notebook to the "all notes" heading
- Creating new notebooks using the "+" button
- Changing focus using the up and down arrow keys

# Screen recordings

## Faster notebook switching

https://github.com/laurent22/joplin/assets/46334387/2d8c53e5-4545-4499-af01-c5a4492670a3

## Faster tagging

In the recording below, the main screen is open in the background in both copies of Joplin. The numbers of notes/tags/notebooks in both profiles are similar to [in the tests done here](https://discourse.joplinapp.org/t/performance-issue/37207/23?u=personalizedrefriger).

https://github.com/laurent22/joplin/assets/46334387/36d4b32f-b0a7-4e63-9170-df6575648cf3
